### PR TITLE
MBS-486: Add support for years BC 

### PR DESCRIPTION
**Pull request by reosarevok**
### Implement MBS-486

The dates are stored in standard astronomical format (1 BCE = 0) but they are always shown to the user as the more intuitive version without a year 0 on the website. On the WS, the astronomical format is retained.

### CODE REVIEW

File: root/utility/age.js [hunk 1]
```diff
@@ -28,7 +28,7 @@ export function hasAge<+T: $ReadOnly<{...DatePeriodRoleT, ...}>>(entity: T): boo
    * If there is no begin year, there is no age.
    * Only compute ages when the begin date is AD.
```
Comment (mwiencek) [inline root/utility/age.js:29]:
```suggestion
```

Also we have some tests for `age` in root/static/scripts/tests/utility.js, would be nice to add some that test negative years

**Review by mwiencek — CHANGES_REQUESTED** [review]
I didn't test this yet, but just a few small things I noticed so far

File: root/static/scripts/edit/utility/dates.js [hunk 2]
```diff
@@ -44,7 +45,9 @@ export const isDateValid = function (y, m, d) {
 };
 
 export const isYearFourDigits = function (y) {
-  return (y === null || y === '' || y.length === 4);
+  return (y === null || y === '' ||
+          y.length === 4 && !isNaN(y.charAt(0)) ||
```
[the remark this answers is not in the document]
Comment (mwiencek) [reply root/static/scripts/edit/utility/dates.js:49]: Sorry, I was wrong because confused that with `Number.isNaN`. Though I think the `RegExp` might be more clear still - right now this is relying on the fact that `'-'` can't be coerced into a number, which is a bit weird.

**Review by mwiencek — APPROVED** [review]
This seems like a good start.

I can't enter "-0001" in the release or relationship editors, but I'm not sure these are useful there (i.e. if it's worth fixing the validation in two more places). Anyway, those cases will likely start working as we consolidate the date period components in the future.

**Comment by reosarevok** [comment]
It might be useful in the relationship editor (but quite rarely) and should probably never happen in the release editor anyway. I *could* add it to the relationship editor, if we want to. We might need to check if it anything needs to be changed in the alias form fields if those get merged sooner.

**Comment by reosarevok** [comment]
Rebased this on current master, seems to actually work fine in the alias form too, so hopefully it's still fine!

**Comment by yvanzo** [comment]
Wouldn’t that break JSON+LD given <https://schema.org/Date>?

**Comment by reosarevok** [comment]
Well, in theory ISO_8601 supports BC years (https://en.wikipedia.org/wiki/ISO_8601 says so at least) and I assume given there's plenty of philosophers and whatnot probably supported by schema.org, they are able to parse those. But I haven't actively tested JSON-LD on these, are you seeing issues, other than the year being negative?

**Comment by yvanzo** [comment]
> Well, in theory ISO_8601 supports BC years (https://en.wikipedia.org/wiki/ISO_8601 says so at least)

Not like that; See [Monkey’s comment](https://tickets.metabrainz.org/browse/MBS-486?focusedCommentId=51098#comment-51098) about it. It requires at least five digits or it can be mistaken with a shortened notation.

(By the way, that comment has been entered during Summit 19; See [related notes](https://docs.google.com/document/d/1abk9f755ewoCBvuoxJPRgpFB0qXeICbNuzcqx6DBhcQ/edit#heading=h.i35fnpvxd49l).

**Comment by reosarevok** [comment]
Hmm. I'm not sure I'm seeing the issue. My understanding is that if it goes over the 4 digits (plus - sign) then we need to have a specified number of extra digits, but the example just after for example mentions that 2BC = −0001. The LOC specs for example have 4 digits by default, it seems: http://www.loc.gov/standards/datetime/

**Comment by reosarevok** [comment]
This now works on relationships, since they use the React components. It doesn't work on the release editor, but given BCE releases are AFAICT not a thing, it's probably fine.

**Comment by mwiencek** [comment]
If I understand the issue correctly, `PartialDateInput` stores dates in BCE format in its state. And I guess we can't change that because other pages using it (edit profile form, alias edit form, external link attribute dialog) have to post BCE dates to the server.

Perhaps we can change `formatDate` to accept an additional argument that specifies what format the input date is in?

**Comment by reosarevok** [comment]
Yeah. I was just not sure how to tell whether the date is new (state-only) or old - how would you keep track in a sane way?
